osu.Server.Spectator/Hubs/MultiplayerHub.cs

```diff
@@ -438,6 +443,32 @@ private async Task changeUserMods(IEnumerable<APIMod> newMods, MultiplayerRoom r
             await Clients.Group(GetGroupId(room.RoomID)).UserModsChanged(CurrentContextUserId, newModList);
         }
 
+        private static void ensureSettingsModsValid(MultiplayerRoomSettings settings)
+        {
+            // check against ruleset
+            if (!populateValidModsForRuleset(settings.RulesetID, settings.RequiredMods, out var requiredMods))
+            {
+                var invalidRequiredAcronyms = string.Join(',', settings.RequiredMods.Where(m => requiredMods.All(valid => valid.Acronym != m.Acronym)).Select(m => m.Acronym));
+                throw new InvalidStateException($"Invalid mods were selected for specified ruleset: {invalidRequiredAcronyms}");
+            }
+
+            if (!populateValidModsForRuleset(settings.RulesetID, settings.AllowedMods, out var allowedMods))
+            {
+                var invalidAllowedAcronyms = string.Join(',', settings.AllowedMods.Where(m => allowedMods.All(valid => valid.Acronym != m.Acronym)).Select(m => m.Acronym));
+                throw new InvalidStateException($"Invalid mods were selected for specified ruleset: {invalidAllowedAcronyms}");
+            }
+
+            if (!ModUtils.CheckCompatibleSet(requiredMods, out var invalid))
+                throw new InvalidStateException($"Invalid combination of required mods: {string.Join(',', invalid.Select(m => m.Acronym))}");
+
+            // check aggregate combinations with each allowed mod individually.
+            foreach (var allowedMod in allowedMods)
+            {
+                if (!ModUtils.CheckCompatibleSet(requiredMods.Concat(new[] { allowedMod }), out invalid))
+                    throw new InvalidStateException($"Invalid combination of required and allowed mods: {string.Join(',', invalid.Select(m => m.Acronym))}");
```

Comment:
Just going to note that there are some game-side cases in which this may run counter to players' expectations. Consider someone wanting to have Sudden Death as a required mod and Perfect as a free mod, which technically kinda makes sense as Perfect is a stricter Sudden Death. With this change that is not permitted anymore. But I'm not sure whether it's a huge deal right now.

Reply:
This is an interesting one for sure. I think we *do* want to allow such cases, but this should probably either be done in `CheckCompatibleSet` or via a new `ModUtils` method, so we keep this logic all in one place.